### ISSUE TITLE
Clarity: refactor the DB to support chain forks

### ISSUE THREAD

**Issue by jcnelson** [blockstack/blockstack-core#1034]
Right now, the Clarity DB indexes persistent variables and data maps in a way that does not admit chain forks, or even tying operations to transactions.  This will need to change at some point, so we can (1) anchor instantiation/update/deletes to the points in the chain history where they occurred, and (2) maintain multiple materialized views of the DB (required for tolerating forks).  To get there, we'll need to do at least the following:

* Use the (block hash, txid) as the primary key for each persisted variable and data map (and their values)
* Don't delete anything -- just mark it as deleted, but keep the value around in case a fork "un-does" the delete

**Comment by kantai** [comment]
> Use the (block hash, txid) as the primary key for each persisted variable and data map (and their values)

This cannot be -- block has and txid are not unique for entries in the database (a single TX may have caused multiple writes). These should be added as non-primary-key, but still probably indexed, fields, though.

> Don't delete anything -- just mark it as deleted, but keep the value around in case a fork "un-does" the delete.

Yep, that's already how it works. Each row in the table is an reverse-able operation. Deletes just store NULLs. Fetches work via doing an indexed query sorted by recency, limited to 1 result.

**Comment by jcnelson** [comment]
> This cannot be -- block hash and txid are not unique for entries in the database (a single TX may have caused multiple writes). These should be added as non-primary-key, but still probably indexed fields, though.

Ah, you're right.  The primary key needs to additionally contain a per-(block-hash, txid) counter to distinguish each operation within the transaction.  Let me try again -- we need to be able to link each mutation to the database to a (block-hash, txid) pair, and within a single (block-hash, txid), identify the sequence of mutations that occurred.  Then, when processing reads while evaluating the code in a transaction, we'll be able to efficiently select which data to return (based on (1) which fork the block containing the transaction containing the code is working on, and (2) where in the block the transaction occurs).

**Comment by kantai** [comment]
I'm still not sure why they'd need to be part of a primary key. The primary key currently is just an auto-incrementing ID, which seems sufficient. All that needs to be done is adding two fields (block-hash and txid) and an index on them.

**Comment by jcnelson** [comment]
Going to go ahead and close this issue in favor of one that calls for integrating the MARF into the database.
